core/services/relay/evm/chain_reader.go

```diff
@@ -164,23 +169,57 @@ func (cr *chainReader) addMethod(
 func (cr *chainReader) addEvent(contractName, eventName string, a abi.ABI, chainReaderDefinition types.ChainReaderDefinition) error {
 	event, eventExists := a.Events[chainReaderDefinition.ChainSpecificName]
 	if !eventExists {
-		return fmt.Errorf("%w: method %s doesn't exist", commontypes.ErrInvalidConfig, chainReaderDefinition.ChainSpecificName)
+		return fmt.Errorf("%w: event %s doesn't exist", commontypes.ErrInvalidConfig, chainReaderDefinition.ChainSpecificName)
 	}
-	cr.contractBindings.AddReadBinding(contractName, eventName, &eventBinding{
-		lp:   cr.lp,
-		hash: event.ID,
-	})
 
-	// Though nothing is encoded encoderDef is required so that CreateType can return a struct{} for CreateType to allow "decoding" into.
-	// The caller isn't aware that there are no arguments and will try to encode the parameters.
-	// The "Arguments" must be empty so decoding is to struct{}, prefix doesn't matter, as this won't be encoded.
-	if err := cr.addEncoderDef(contractName, eventName, abi.Arguments{}, nil, chainReaderDefinition); err != nil {
+	filterArgs, topicInfo, indexArgNames := setupEventInput(event, chainReaderDefinition)
+
+	if err := verifyEventInputsUsed(chainReaderDefinition, indexArgNames); err != nil {
 		return err
 	}
 
+	if err := topicInfo.Init(); err != nil {
+		return err
+	}
+
+	// Encoder def's codec won't be used to encode, only for its type as input for GetLatestValue
+	if err := cr.addEncoderDef(contractName, eventName, filterArgs, nil, chainReaderDefinition); err != nil {
+		return err
+	}
+
+	ce := cr.parsed.encoderDefs[WrapItemType(contractName, eventName, true)]
+	inMod, err := chainReaderDefinition.InputModifications.ToModifier(evmDecoderHooks...)
+	if err != nil {
+		return err
+	}
+
+	if _, err = inMod.RetypeForOffChain(reflect.PointerTo(ce.checkedType), ""); err != nil {
+		return err
+	}
```

Comment:
This is basically using `addEncoderDef` to get the `checkedType`, then adds input modifications to it to use it for the `GetLatestValue` args? This whole block doesn't sit right with me, maybe it would be better off as a separate function that just gets the checkedType with the modifications?